app/models/manageiq/providers/ansible_tower/shared/inventory/parser/automation_manager.rb

```diff
@@ -122,7 +122,7 @@ def credentials
     collector.credentials.each do |credential|
       inventory_object = persister.credentials.find_or_build(credential.id.to_s)
       inventory_object.name = credential.name
-      inventory_object.userid = credential.username
+      inventory_object.userid = credential.username || "<no username provided>"
```

Comment:
@jeffwarnica would it be better to check if userid is nil before assuming it is present rather than set it to a "random" string?  Doing this turns a simple `unless userid.nil?` into a string compare, and actually sending `"<no username provided>"` to a provider is probably not what we want.

Reply:
If find_or_build() found something, then we are updating a record. The LHS, if once a string, is invalid; "<no username provided>" is more correct then the username that someone removed from Tower.

`|| nil` might also be OK, but since I've hit an unrelated UI problems displaying `nil` today, caution, in giving it something, seems reasonable here.

Reply:
> If find_or_build() found something, then we are updating a record.

Mmm that's not quite right, find_or_build looks in the inventory_collections not the database so it could "find" an inventory object that was created by another collection's parser.  Either way I don't think it is relevant.

> The LHS, if once a string, is invalid "" is more correct then the username that someone removed from Tower.

If someone removed a username from tower and `credential.username == nil` then isn't setting `inventory_object.userid = nil` correct?

> || nil might also be OK

`|| nil` isn't needed, if `credential.username || "<...>"` would work then `credential.username` is already `nil` and just `inventory_object.userid = credential.username` would work.

> but since I've hit an unrelated UI problems displaying nil today, caution, in giving it something, seems reasonable here.

Do you have a traceback from where the userid being `nil` caused an error?  It is probably not hard to fix that.  If we have issues with this then we might need a data migration to update any nil userid's to some other string.

Unless we have presence validations on the model (we don't) then nil here should be a valid value.

Reply:
Just noticed you added a BZ, looking at the logs the error is:
```
ERROR -- : [NoMethodError]: undefined method `username' for #<AnsibleTowerClient::Credential:0x0000000009ba3098>  Method:[block in method_missing]
ERROR -- : /opt/rh/cfme-gemset/bundler/gems/cfme-providers-ansible_tower-9841e4e2793c/app/models/manageiq/providers/ansible_tower/shared/inventory/parser/automation_manager.rb:101:in `block in credentials'
```

Which means the issue isn't that `credential.username` is nil, it means that `credential` doesn't even respond_to `username` so you would need something like `inventory_object.userid = credential.username if credential.respond_to?(:username)`

Reply:
My solution does work, but I'll defer to any better rubyist around if there is a more correct solution.

Doesn't guard against something that wants inventory_object.userid to be non-nill later, though. 

Which ever you want.

Reply:
> My solution does work

That's interesting, I tried that out and I get the same error:
```
>> AnsibleTowerClient::Credential.new('/api', {}).username || "no credentails"
NoMethodError: undefined method `username' for <AnsibleTowerClient::Credential >:AnsibleTowerClient::Credential
```

which is expected because `||` doesn't handle a `NoMethodError` exception earlier in the statement.